REF: Internal / External values

TomAugspurger opened the pull request:
Closes #19548

Just changed Index so far. Need to

- Run an ASV
- add `Series._ndarray_values`
- Start cleaning up now redundant if conditions.

TomAugspurger committed:
REF/Clean: Internal / External values

TomAugspurger commented on doc/source/internals.rst:
```diff
@@ -89,6 +89,21 @@ not check (or care) whether the levels themselves are sorted. Fortunately, the
+  always an ``ndarray`` or ``ExtensionArray``).
+
+So, for example, ``Series[category]._values`` is a ``Categorical``, while ``Series[category]._ndarray_values`` is
+the underlying ndarray.
```
Not sure what `Series[categorical]._ndarray_values` should be, categories or codes? `PeriodIndex._ndarray_values` is the ordinals, so maybe codes?

jreback replied:
yes

TomAugspurger committed:
Move to index base

jreback commented on doc/source/internals.rst:
```diff
@@ -89,6 +89,21 @@ not check (or care) whether the levels themselves are sorted. Fortunately, the
+Values
+~~~~~~
+
+Pandas extends NumPy's type system in a few places, so we have multiple notions of "values" floating around.
```
The first sentence is totally not clear to a new reader

jreback commented on doc/source/internals.rst:
```diff
@@ -89,6 +89,21 @@ not check (or care) whether the levels themselves are sorted. Fortunately, the
+
+* ``cls._ndarray_values`` is *always* and ``ndarray``
+* ``cls._values`` refers is the "best possible" array. This could be an ``ndarray``, ``ExtensionArray``, or
+  in ``Index`` subclass (note: we're in the process of removing the index subclasses here so that it's
```
you are using internal (pun intended) jargon here

jreback commented on pandas/core/base.py:
```diff
@@ -973,8 +999,12 @@ def value_counts(self, normalize=False, sort=True, ascending=False,
     def unique(self):
         values = self._values
-
+        if isinstance(values, ABCDatetimeIndex):
```
why are you special casing?

jreback commented on pandas/core/dtypes/concat.py:
```diff
@@ -498,9 +498,16 @@ def _concat_index_asobject(to_concat, name=None):
-    return self._shallow_copy_with_infer(np.concatenate(to_concat), **attribs)
+    arrays = []
+    for x in to_concat:
+        if is_categorical_dtype(x):
```
why is this a special case?

jorisvandenbossche replied:
@TomAugspurger don't we want ``np.asarray(x, dtype=object)`` for *all* extension dtypes in this case? 

So for Period and Interval `_values`  (used some lines below for Index types) currently still returns object array, but I think the purpose is that those will also return their own Array class, so using ``np.asarray(x, dtype=object)`` will do the same now and is more future proof?

jreback commented on pandas/core/indexes/base.py:
```diff
@@ -2208,27 +2208,37 @@ def union(self, other):
             other = other.astype('O')
             return this.union(other)
 
+        if is_categorical_dtype(self):
```
why are these special caes?

jreback reviewed (changes requested):
so a) this should remove .get_values() as well completely, adding something w/o removing the old to me is a non-starter here.

I see lots of special cases. This points to a problem in the interface. if you are doing is_......_ then it needs to be fixed in ExtensenionArray (IOW have a dispatch for that function) and overrides in the subclasses. calling out Categorical here.

TomAugspurger committed:
Cleanup unique handling

TomAugspurger committed:
Merge remote-tracking branch 'upstream/master' into index-values

TomAugspurger committed:
Simplify object concat

TomAugspurger committed:
Use values for intersection

I think eventually we'll want to ndarray_values for this, but it'll
require a bit more work to support. Currently, using ndarary_values
causes occasional failures on categorical.

TomAugspurger committed:
hmm

TomAugspurger committed:
Merge remote-tracking branch 'upstream/master' into index-values

TomAugspurger committed:
Additional testing

TomAugspurger committed:
More tests

TomAugspurger committed:
ndarray_values

TomAugspurger committed:
API: Default ExtensionArray.astype

(cherry picked from commit 943a915562b72bed147c857de927afa0daf31c1a)
(cherry picked from commit fbf0a0672380e210d3cb3c527fa8045a204d81be)

TomAugspurger committed:
Simplify concat_as_object

TomAugspurger committed:
Py2 compat

(cherry picked from commit b20e12cae68dd86ff51597464045656763d369f7)

TomAugspurger committed:
Set-ops ugliness

TomAugspurger committed:
better docstrings

TomAugspurger committed:
tolist

pep8speaks commented:
Hello @TomAugspurger! Thanks for updating the PR.













 - In the file [`pandas/core/indexes/datetimes.py`](https://github.com/pandas-dev/pandas/blob/b012c1967b6de548b999514fe4b560ba9b7ee635/pandas/core/indexes/datetimes.py), following are the PEP8 issues :

> [Line 1108:80](https://github.com/pandas-dev/pandas/blob/b012c1967b6de548b999514fe4b560ba9b7ee635/pandas/core/indexes/datetimes.py#L1108): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (83 > 79 characters)

TomAugspurger committed:
linting

TomAugspurger commented:
Still a WIP, but this should hopefully pass.

I'm going to make a pass through the library to remove things that are currently doing checks for `Series` vs. `Index`, which should clean things up.

I've also cherry picked `ExtensionArray.astype` stuff from https://github.com/pandas-dev/pandas/pull/19604, so the diff in lines in this PR is a bit smaller than it looks.

Setops are even more of a mess now. I have a biggish refactor started (with nice perf boosts), but didn't finish it today. It's tricky since the current implementation was relying on `self._inner_indexer(self._values, ...)` raising in some cases that now won't raise (period or categorical I think). I think that it's OK to accept a bit of technical debt in this case, and I'll pay it down later.

jorisvandenbossche commented on pandas/core/indexes/base.py:
```diff
@@ -594,6 +596,39 @@ def values(self):
+        index             | values          | _values     | _ndarray_values |
+        ----------------- | -------------- -| ----------- | --------------- |
+        CategoricalIndex  | Categorical     | Categorical | codes           |
+        DatetimeIndex[tz] | ndarray[M8ns]   | DTI[tz]     | datetime@UTC    |
```
datetime@UTC  -> is the same as ndarray[M8ns] in the 'values' column ?

jorisvandenbossche commented on pandas/core/indexes/base.py:
```diff
@@ -594,6 +596,39 @@ def values(self):
+        index             | values          | _values     | ndarray_values |
+        ----------------- | --------------- | ----------- | -------------- |
+        PeriodIndex       | ndarray[object] | PeriodArray
+        IntervalIndex     | IntervalArray   | ndarray[Interval]
```
This should be the other way around? 
and for PeriodIndex, `_ndarray_values` is ndarrary[ordinals] ? (for Interval there is not really a native ndarray representation possible, so that is also object array ?

jorisvandenbossche commented on pandas/core/indexes/category.py:
```diff
@@ -293,6 +304,23 @@ def values(self):
 
+    @property
+    def _values(self):
+        return self._data
```
Is there a difference between `._data` and `.values` ? As below you use `.values` for itemsize and nbytes, but both seems to point to the underlying Categorical

TomAugspurger replied:
Probably just me being inconsistent.

jorisvandenbossche commented on pandas/core/indexes/category.py:
```diff
@@ -769,7 +797,7 @@ def _evaluate_compare(self, other):
     def _delegate_method(self, name, *args, **kwargs):
         """ method delegation to the ._values """
-        method = getattr(self._values, name)
+        method = getattr(self.values, name)
```
Why change this?

TomAugspurger replied:
Seems to be unnecessary. Reverting.

jorisvandenbossche commented on pandas/core/indexes/multi.py:
```diff
@@ -799,9 +799,11 @@ def values(self):
             if box and len(lev) > len(lab):
-                taken = lev._box_values(algos.take_1d(lev._values, lab))
+                taken = lev._box_values(algos.take_1d(lev._values,
+                                                      lab))
```
don't the algos functions always assume ndarray?

TomAugspurger replied:
Yes, fixed (the only one that should be `._values` is the one on line 809 below).

jorisvandenbossche commented on pandas/tests/test_base.py:
```diff
@@ -1175,3 +1176,94 @@ def test_iter_box(self):
+    elif isinstance(l_values, pd.Index):
+        tm.assert_index_equal(l_values, r_values)
+    elif pd.api.types.is_categorical(l_values):
+        tm.assert_categorical_equal(l_values, r_values)
```
In the future, we should maybe add an `assert_array_equal` that handles the different extension arrays as well, then such a special case should not be needed here?

jorisvandenbossche commented on pandas/tests/test_base.py:
```diff
@@ -1175,3 +1176,94 @@ def test_iter_box(self):
+    assert r_values.dtype == dtype
+
+
+def test_values_periodindex():
```
this is also tested in the test above?

jorisvandenbossche reviewed (commented):
> this should remove .get_values() as well completely, adding something w/o removing the old to me is a non-starter here.

@jreback if you are speaking about Series.get_values and Index.get_values, then I don't think this related. Yes, ideally we can implement `get_values` cleanly with one of the other ones, but `get_values` is a public function that should not just be removed or changed. We can certainly discuss deprecating it, but IMO that is separate of this *internal* cleaning up.

jreback commented:
``.get_values()`` should be deprecated, if you want to implement in terms of an internal function, ok. but the external API needs fixing. This was never meant to be exposed and is useless. let's clean that up first

jorisvandenbossche commented:
> external API needs fixing

What needs to be fixed? Or do just mean deprecating it?

>  let's clean that up first

That doesn't need to happen first, it can be done independently of this. I will open an issue for it.

jorisvandenbossche commented:
You were faster :-) https://github.com/pandas-dev/pandas/issues/19617

TomAugspurger committed:
Moved dtypes

(cherry picked from commit d1362271bca8a7b183f3241e5c2f040c422118b8)

TomAugspurger committed:
clean

TomAugspurger committed:
cleanup

TomAugspurger committed:
NumPy compat

TomAugspurger committed:
Use base _values for CategoricalIndex

TomAugspurger committed:
Update dev docs

TomAugspurger committed:
Merge remote-tracking branch 'upstream/master' into index-values

TomAugspurger committed:
cleanup

TomAugspurger committed:
Merge remote-tracking branch 'upstream/master' into index-values

TomAugspurger committed:
Linting

TomAugspurger committed:
Precision in tests

TomAugspurger committed:
Merge remote-tracking branch 'upstream/master' into index-values

TomAugspurger commented:
Latest push includes the https://github.com/pandas-dev/pandas/pull/19558. Just making sure that CI passes, the other PR will have to be merged first.

jreback commented on doc/source/internals.rst:
```diff
@@ -89,6 +89,25 @@ not check (or care) whether the levels themselves are sorted. Fortunately, the
 constructors ``from_tuples`` and ``from_arrays`` ensure that this is true, but
 if you compute the levels and labels yourself, please be careful.
 
+Values
```
you could add section tags

jreback commented on pandas/core/base.py:
```diff
@@ -768,8 +769,21 @@ def base(self):
+        See '_values' for more.
+        """
+        # type: () -> np.ndarray
+        from pandas.core.dtypes.common import is_categorical_dtype
```
this does NOT belong here. you *already* have a sub-clss EA for Categorical that can simply override this.

jorisvandenbossche replied:
This code is to return `_ndarray_values` for both Series and Index, and for Series there is in general no subclass that can override it. 
Of course we could say to put this logic in the Blocks, but that is then also not for Index, so not sure it is better.

This raises the question for me, though, what this will return for external extension types. Since it is is `.values`, I suppose this is a materialized array?

TomAugspurger replied:
I've moved it all the way to `ExtensionArray`, with the default being just returning an `np.array(self)`, same as before.  It's not part of the interface though. I think is what @jorisvandenbossche  suggested [here](https://github.com/pandas-dev/pandas/pull/19558/files#r167408753).

jreback commented on pandas/core/base.py:
```diff
@@ -768,8 +769,21 @@ def base(self):
+        # type: () -> np.ndarray
+        from pandas.core.dtypes.common import is_categorical_dtype
+
+        if is_categorical_dtype(self):
```
why do we have this at all, e.g. ``np.array(ea)`` should just do the right thing no? why are we adding something else? note I actually don't mind that we have an additional property that we use consistently, more to the point of why ``__array__`` does not just return ``_ndarray_values``

jreback commented on pandas/core/base.py:
```diff
@@ -819,8 +833,10 @@ def tolist(self):
 
         if is_datetimelike(self):
```
this should be overriden in EA, rather than specific dispatching via if/else here, IOW it should be a part of the interface, or be defined as ``list(.values)``

jorisvandenbossche replied:
I am not sure it is needed to add a `tolist` to the interface, I think in general it can rely on `__iter__`/`__getitem__` (so `list(self._values)`). 
The only problem in that case is that we still need to distinguish between normal numeric types (where the above would return numpy scalars and not python scalars) and other types where this already gives the correct result.

jreback commented on pandas/core/indexes/base.py:
```diff
@@ -2228,27 +2264,37 @@ def union(self, other):
             other = other.astype('O')
             return this.union(other)
 
+        # TODO: setops-refactor, clean all this up
```
this if/then needs to all be completely removed

TomAugspurger replied:
Can't be removed until setops are fix more generally. [Like I said](https://github.com/pandas-dev/pandas/pull/19558#issuecomment-364280424), I have a PR started, but it's a bigger issue than what's on the critical path for ExtensionArray.

Take `CategoricalIndex.intersection` We *relied* on `CategoricalIndex._outer_indexer(self._values, other._values)` raising a TypeError to get the correct result. If I pass `CategoricalIndex._ndarray_values`, we don't get a type error any more, but we get the wrong result.

There's no way to avoid the if statements until set ops are completely refactored.

jreback replied:
ok, let's do this next then. these if/then/else are really smelly.

jreback replied:
if *might* be worth it to put this in a single method now (not sure if its possible) to isolate this code, maybe

```
def _get_internal_type(other):
     if is_period_dtype(other) ......
          ...
    else:
         ...
```

jreback commented on pandas/core/indexes/base.py:
```diff
@@ -2311,20 +2357,30 @@ def intersection(self, other):
 
+        # TODO: setops-refactor, clean all this up
+        if is_period_dtype(self):
+            lvals = self._ndarray_values
```
same

jreback commented on pandas/core/indexes/datetimes.py:
```diff
@@ -1086,6 +1095,19 @@ def snap(self, freq='S'):
+        # So we extract the tz-naive DatetimeIndex, unique that, and wrap the
+        # result with out TZ.
+        if self.tz is not None:
+            naive = type(self)(self._ndarray_values, copy=False)
```
huh?

jorisvandenbossche replied:
Do you have a concrete question? It's not the nicest design, but for me the comment above is clear (and I suppose the idea is that this is temporary until there is a DatetimeTZArray)

jreback replied:
all of this hacking around is not great. The problem is these temporary work-around become more and more. Really need to bite the bullet and fix up the implementations to avoid this.

jreback commented on pandas/core/indexes/multi.py:
```diff
@@ -2655,8 +2656,9 @@ def equals(self, other):
-            ovalues = algos.take_nd(np.asarray(other.levels[i]._values),
-                                    olabels, allow_fill=False)
+            ovalues = algos.take_nd(
+                np.asarray(other.levels[i]._values),
```
why not _ndarray_values here?

TomAugspurger replied:
Breaks for categoricals. We've baked knowledge of cats into `algost.take_nd` / `take_1d`, to handle it.

jreback commented on pandas/tests/test_base.py:
```diff
@@ -1175,3 +1176,136 @@ def test_iter_box(self):
             assert res == exp
+
+
+@pytest.mark.parametrize('arr, expected', [
```
this is not the right location for these. they need to go in appropriate test locations.

jreback reviewed (changes requested):
this needs work

TomAugspurger commented:
Are indexes inheriting from EA the plan, or will we do composition like CategoricalIndex composes a Categorical. I think I like composition a bit more, but both have their merits

________________________________
From: Jeff Reback <notifications@github.com>
Sent: Saturday, February 10, 2018 9:46:06 AM
To: pandas-dev/pandas
Cc: Tom Augspurger; Mention
Subject: Re: [pandas-dev/pandas] [WIP]REF: Internal / External values (#19558)


I think you might want to inherit the Indexes from EA before you attempt to do this PR.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/pandas-dev/pandas/pull/19558#issuecomment-364664191>, or mute the thread<https://github.com/notifications/unsubscribe-auth/ABQHIm_4ApivlYtdN1d_sv8vvnO9V82kks5tTbm-gaJpZM4R7rd_>.

jreback commented:
Indexes are perfect candidates for EA.

jorisvandenbossche commented:
I didn't think about it in detail or looked at the code in detail, but, I think composition would make more sense. Series will not go the inheritance way but will do composition, so doing the same for Index will keep those two more in line. 
And, it is also more in line to how it currently already is for the indexes based on extension dtypes (period, categorical), so this will be a less big change.

jbrockmendel commented:
> Are indexes inheriting from EA the plan, or will we do composition like CategoricalIndex composes a Categorical. I think I like composition a bit more, but both have their merits

+1 for composition.  

> ._values is either an ndarray, ExtensionArray, or Index (for `DatetimeIndex w/ TZ)
._ndarray_values is always an ndarray

Double-checking I understand the promise/contract here: can I think of `obj._ndarray_values` as "unwrap `obj` until we get to a non-pandas [usually numpy] object; may be lossy" and `obj._values` as "unwrap `obj` as much as we can without being lossy"?

More specifically, will it always be the case that `op(obj, other)` can wrap `op(obj._values, other)` for supported (arith, comparison, [...?]) ops?

Is `_ndarray_values` specifically `np.ndarray` or is this being left open for e.g. pyarrow in the future or geopandas/dask to get creative with?  If the latter, is there a requirement on what the returned object should look like?

jorisvandenbossche commented on pandas/core/indexes/base.py:
```diff
@@ -594,6 +596,40 @@ def values(self):
+
+        index             | values          | _values      | _ndarray_values |
+        ----------------- | --------------- | ------------ | --------------- |
+        PeriodIndex       | ndarray[object] | ndarray[obj] | ndarray[int]    |
```
For the `_values`, shouldn't this be `PeriodIndex` for now? Similar as for DatetimeTZ now has DTI[tz], and so in the future it would become PeriodArray. 

(but not sure where this is currently actually used)

TomAugspurger replied:
No I don't think so, since `Series[Period]._values` returns an ndarray of objects. `Series[datetime-with-TZ]._values` is the only special one, and I'm only changing `Index._values` to match `Series._values` for now.

jorisvandenbossche replied:
Yes, but that is because we don't have a PeriodBlock that is backed by a PeriodIndex, as is the case for DatetimeTZ, so at the moment you cannot really have Period values in a series. 
I am not saying that this should be the case now, only that I think it might be more logical. But, it might well be that we then first need to add a working "PeriodBlock" (ExtensionBlock backed by PeriodIndex/PeriodArray) to have this work.

TomAugspurger replied:
I'm confused what your proposing / saying then. Is it just that `._values` being a `PeriodIndex` would be more consistent with `._values` returning a `DatetimeIndex`?

What we have now "works", but is internally inconsistent. I'm updating `Index._values` to be consistent with `Series._values`, and changing uses of `.values` / `._values` / `._ndarray_values` to be the correct one for that use.

jorisvandenbossche replied:
> Is it just that ._values being a PeriodIndex would be more consistent with ._values returning a DatetimeIndex?

Yes. So regardless of what `._values` currently does or doesn't return for Index/Series. Just from looking at your table, returning `PeriodIndex` would make more sense IMO, as this is (at this moment) the closest thing to an array-like that preserves the information. 
But, I don't know what implication it would have for Series ops that `._values` would start returning a PeriodIndex instead of object array of Periods, so it was really just a question, not a request for change.

TomAugspurger committed:
Push _ndarray_values to ExtensionArray

Now IndexOpsMixin._ndarray_values will dispatch all the way down to the EA.
Subclasses like Categorical can override it as they see fit.

TomAugspurger committed:
Clean up tolist

TomAugspurger committed:
Move test locations

TomAugspurger commented:
Gotcha. I'm not sure either, but it'd complicate things I suspect, since we
don't have a PeriodBlock like we do for DatetimeTZ.

Regardless, this will change in the near future to be a PeriodArray.



On Sun, Feb 11, 2018 at 2:50 PM, Joris Van den Bossche <
notifications@github.com> wrote:

> *@jorisvandenbossche* commented on this pull request.
> ------------------------------
>
> In pandas/core/indexes/base.py
> <https://github.com/pandas-dev/pandas/pull/19558#discussion_r167449141>:
>
> > +        Both ``_values`` and ``_ndarray_values`` are consistent between
> +        ``Series`` and ``Index``.
> +
> +        It may differ from the public '.values' method.
> +
> +        index             | values          | _values     | _ndarray_values |
> +        ----------------- | -------------- -| ----------- | --------------- |
> +        CategoricalIndex  | Categorical     | Categorical | codes           |
> +        DatetimeIndex[tz] | ndarray[M8ns]   | DTI[tz]     | ndarray[M8ns]   |
> +
> +        For the following, the ``._values`` is currently ``ndarray[object]``,
> +        but will soon be an ``ExtensionArray``
> +
> +        index             | values          | _values      | _ndarray_values |
> +        ----------------- | --------------- | ------------ | --------------- |
> +        PeriodIndex       | ndarray[object] | ndarray[obj] | ndarray[int]    |
>
> Is it just that ._values being a PeriodIndex would be more consistent with
> ._values returning a DatetimeIndex?
>
> Yes. So regardless of what ._values currently does or doesn't return for
> Index/Series. Just from looking at your table, returning PeriodIndex
> would make more sense IMO, as this is (at this moment) the closest thing to
> an array-like that preserves the information.
> But, I don't know what implication it would have for Series ops that
> ._values would start returning a PeriodIndex instead of object array of
> Periods, so it was really just a question, not a request for change.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/pandas-dev/pandas/pull/19558#discussion_r167449141>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABQHIphuJbL5OS7V268SM_K_V_b-5Ux4ks5tT1KWgaJpZM4R7rd_>
> .
>

TomAugspurger committed:
Fixed test

jorisvandenbossche commented on pandas/core/dtypes/concat.py:
```diff
@@ -480,20 +480,22 @@ def _concat_datetimetz(to_concat, name=None):
 def _concat_index_same_dtype(indexes, klass=None):
     klass = klass if klass is not None else indexes[0].__class__
-    return klass(np.concatenate([x._values for x in indexes]))
+    return klass(np.concatenate([x._ndarray_values for x in indexes]))
```
This one is only used for numeric indices, so `_values` or `_ndarray_values` should not matter. 
I was mainly thinking that for those cases where `_ndarray_values` actually differs from `_values`, like a categorical, the above code will not work anyway, as `klass(codes)` is not enough to reconstruct the categorical. So `_values` seems safer to me.

jorisvandenbossche commented on pandas/core/arrays/base.py:
```diff
@@ -266,3 +266,15 @@ def _can_hold_na(self):
+        # type: () -> np.ndarray
+        """Internal pandas method for lossy conversion to a NumPy ndarray.
+
+        This method is not part of the pandas interface.
```
I find it a bit strange that we say this is not part of the interface, but still provide here a default implementation and say what it is. 

If it is not part of the interface, we could also define it on our own subclasses without defining it here. 
But on the other hand, that will raise errors when an external extension array does not have this method. Eg some attributes on Series will call into this, which makes it somehow part of the interface.

TomAugspurger replied:
> I find it a bit strange that we say this is not part of the interface, but still provide here a default implementation and say what it is.

This simplifies the implementation since

a.) we don't have to have a `PandasExtensionArray` that subclasses `ExtensionArray`, just to define this method.
b.) we can safely call `._values` anywhere in our code without checking whether it's an extension array or a pandas extension array.

My preference is to leave it out of the interface until someone sees an actual need for it. I suspect this could come up if / when we start allowing custom indexes with their own indexing engines, but that seems like a ways down the road...

jorisvandenbossche replied:
Yes, I understand those reasons, but that means it *is* part of the interface (in the sense that if somebody would be stupid to implement a `_ndarray_values` property that returns something differently, it will break code?) 
It's just a part of the interface that we ask not to implement?

To give an example, for GeoPandas, I was wondering if I would overwrite this property to return my `self.data` (the integer pointers) instead of a materialized array (which can be costly). 
For things like `Series.strides`, `Series.itemsize`, .. this will be OK (and in principle no user should call this anyway .. so maybe I should not worry about it), but if it is used in other places as well that I am not really sure about, such an implementation might actually give a wrong result (but at the same time, if this is called in certain places for my column with an extension array, I want to know that, because materializing can be costly and I want to avoid this as much as possible).

TomAugspurger replied:
I see your point. Your GeoPandas concern is a good one. But I'm not sure how to proceeded :/ Do you have a preference for

a.) Adding it to the interface
b.) Leaving it out of the interface, with a default implementation on EA
c.) Making a pandas-specific EA that defines `_ndarray_values`, so uses of `_ndarray_values` will need to check for that attr.

`._values` / `._ndarray_values` are meant to be internal, and I don't want to limit future refactorings with backwards compatibility concerns. I'm also not even sure how to document it beyond what's already there.

jorisvandenbossche replied:
Maybe the best option is to leave it for now? :)

Ideally, the Series machinery should not use `_ndarray_values` too often, as this would require knowledge of what those values mean (codes, ordinals, ...), which Series does not need to know (if there is knowledge required, it should somehow dispatch the operation to the underlying (extension) array itself). 

And this is already the case I think. From a quick scan, currently in the PR, I see `ndarray_values` is mainly used for:

- indexing (values used in the engine) -> for now we don't support extension arrays to be the data for Index (apart from our own), so this is a worry for later if we want to tackle that
- Index object operations (eg the setops) -> same as above
- in some specific cases where we know the parent object (eg we know it is a PeriodIndex, we just want to get the ordinals)
- for algos (eg take, unique) -> either they dispatch to EA (eg take) or for now the fallback is object array anyhow (eg value_counts), and in general this is something we still need to discuss how to handle in general (factorize, unique, ..)
- some Series attributes (shape, itemsize, strides, size) -> shape can also be derived from `_values` directly (like `nbytes`) as it is part of the EA interface. And from the others, I think `size` would be nice to not have to materialize the EA (itemsize and strides I don't care about). Can we let this return `len(self._values)`, which is part of the interface? (I am not sure to what extent `.size` and `__len__` on numpy arrays are different / have different performance)

TomAugspurger replied:
Your assessment looks correct. Nothing that's series specific currently uses `_ndarray_values`, so it's really just code that is Index-specific or index / series agnostic.

I can change `.shape` and `.size` to use `._values` instead of  instead of `._ndarary` values. I'll just have to override it in `DatetimeIndex` to avoid a recursion error.

jorisvandenbossche commented on pandas/core/base.py:
```diff
@@ -738,22 +740,22 @@ def data(self):
     def nbytes(self):
         """ return the number of bytes in the underlying data """
-        return self._values.nbytes
+        return self._ndarray_values.nbytes
```
Should `nbytes` be called on `self._values` instead? (as we require a `nbytes` implementation on the Extension Array, and numpy arrays already have it)

TomAugspurger replied:
I think this caused issues for CI, but re-running tests now with this change.

jorisvandenbossche commented on pandas/core/indexes/category.py:
```diff
@@ -293,6 +293,15 @@ def values(self):
 
+    @property
+    def itemsize(self):
+        return self.values.itemsize
```
let's add a comment why you are overwriting this here (default is to use `_ndarray_values`, which are codes, but Categorical defines itemsize as the itemsize of the categories). 
Also, would it be more consistent to do `self._values.itemsize`? (not that it would differ in this case)

TomAugspurger replied:
More consistent with the parent's implementation, yes, but I'd prefer to use the public API where possible.

jorisvandenbossche replied:
> , but I'd prefer to use the public API where possible.

But the public api is not meant to be consistent, but to maintain backwards compatibility :-) (at least for now)

jorisvandenbossche commented on pandas/core/indexes/multi.py:
```diff
@@ -1317,7 +1319,7 @@ def from_tuples(cls, tuples, sortorder=None, names=None):
         elif isinstance(tuples, (np.ndarray, Index)):
             if isinstance(tuples, Index):
-                tuples = tuples._values
+                tuples = tuples._ndarray_values
```
Why `ndarray_values` in this case? Because I think this code path can only work if it is an object index that contains tuples, so should not be different in this case.

Although that maybe raises the question: in such cases, where we know both `_values` and `_ndarray_values` will return the same ndarray (eg in all code only for numeric indexes), which one of the two do we take? (might be good to be consistent on this)

jorisvandenbossche commented on pandas/io/formats/format.py:
```diff
@@ -1897,7 +1897,7 @@ def _format(x):
         vals = self.values
         if isinstance(vals, Index):
-            vals = vals._values
+            vals = vals._ndarray_values
```
Currently, an extension array will end up here (using the GenericArrayFormatter), so should this rather be `np.array(vals._values)` ? 
(the implementation of `_ndarray_values` on an ExtensionArray should not influence its formatting?)

TomAugspurger replied:
It seems like by this point we always have a regular `Index`, not an extension array, so it doesn't matter in terms of output. Changing to `_values` since that's clearer.

jorisvandenbossche commented on pandas/tests/indexes/datetimes/test_datetime.py:
```diff
@@ -469,3 +469,18 @@ def test_factorize_dst(self):
+        result = arr.unique()
+
+        if isinstance(expected, np.ndarray):
+            tm.assert_numpy_array_equal(result, expected)
```
The expected is always a DatetimeIndex in the above parametrization

jorisvandenbossche commented on pandas/tests/test_base.py:
```diff
@@ -1175,3 +1176,54 @@ def test_iter_box(self):
+    (pd.DatetimeIndex(['2017-01-01T00:00:00'], tz="US/Eastern"),
+     np.array(['2017-01-01T05:00:00'], dtype='M8[ns]')),
+    (pd.TimedeltaIndex([10**10]), np.array([10**10], dtype='m8[ns]')),
+    pytest.mark.xfail(reason='PeriodArray not implemented')((
```
Why fail? As I think `_ndarray_values` already are the ordinals, and will stay that way? (it is `_values` that will change from object array to PeriodArray?)

TomAugspurger replied:
It's an `xfail` since `Series[period]._ndarray_values` will return an ndarray[Period], while `PeriodIndex._ndarray_values` returns the ordinal.

Passing this now would require modifying `Series._ndarray_values`, which is currently just `self._internal_values`, to special case object dtype to do inference to see if we have periods. I don't think that's worth it since Periods will (hopefully) be a first-class extension dtype by the next release.

jorisvandenbossche replied:
Ah missed that the test below tested it for both Series and Index.

jorisvandenbossche reviewed (commented):
Added some comments, but in general this is looking good to me.

For Index, we now also have the `_data` attribute. Is it the idea that in the future this will become the same as `_values` ?

TomAugspurger commented:
`._data` seems even more inconsistent

class             | output
----------------- | ------
Index             | ndarray
PeriodIndex       | ndarray[int] (ordinals)
DatetimeIndex     | DatetimeIndex (?)
DatetimeIndex[tz] | ndarray[datetime64[ns]]
CategoricalIndex  | Categorical
IntervalIndex     | None

It hopefully wouldn't be much work to convert uses of `._data` to `._values`, but I don't have any plans to do that right now. I'll open up a followup issue though.

TomAugspurger committed:
REF: Update per comments

TomAugspurger commented:
Thanks for the review. I implemented your suggestions where I didn't comment in reply, and tests pass locally.

jorisvandenbossche commented:
> It hopefully wouldn't be much work to convert uses of ._data to ._values, but I don't have any plans to do that right now. I'll open up a followup issue though. #19658

\+ 1, that only just got me thinking. If you create an index from an extension array (directly, or eg by doing `set_index` on a column backed by extension block with extension array), I suppose this creates an object Index (with array of scalars) right now? Might be good to add a test for that? (but that probably belongs in the other PR that implements storing an EA in Series)

TomAugspurger commented:
> I suppose this creates an object Index (with array of scalars) right now? Might be good to add a test for that? (but that probably belongs in the other PR that implements storing an EA in Series)

Yes, and agreed (we hit this indirectly though `._value_counts`), but it'd be good to have an explicit test.

TomAugspurger committed:
lint

TomAugspurger committed:
REF: Use _values for size and shape

TomAugspurger committed:
PERF: Implement size, shape for IntervalIndex

TomAugspurger committed:
PERF: Avoid materializing values for PeriodIndex shape, size

TomAugspurger commented:
Implemented the changes dicussed in https://github.com/pandas-dev/pandas/pull/19558#discussion_r167650514: `.shape` and `.size` use `self._values`.

I implemented those on `PeriodIndex` and `DatetimeIndex`, else we would have materialized the array of objects (which is what we do on master), so we get some performance gains there I guess

master:

```python
In [6]: idx = pd.interval_range(0, 100_000)

In [8]: %timeit idx.size
308 ns ± 9.8 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)
```

pr:

```python
In [3]: %timeit idx.size
1.05 µs ± 9.63 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)
```

jorisvandenbossche commented:
do 'master' and 'pr' need to be swapped?

TomAugspurger commented:
Not sure what happened on that timing I posted earlier. I can't reproduce it.

```python
In [1]: import pandas as pd

In [2]: idx = pd.interval_range(0, 100_000)

In [3]: %time idx.shape
CPU times: user 177 ms, sys: 10.1 ms, total: 188 ms
Wall time: 187 ms
Out[3]: (100000,)
```

```python
In [1]: import pandas as pd

In [2]: idx = pd.interval_range(0, 100_000)

In [3]: %time idx.shape
CPU times: user 18 µs, sys: 2 µs, total: 20 µs
Wall time: 22.9 µs
Out[3]: (100000,)
```

TomAugspurger commented:
CI is green again.

jorisvandenbossche commented on doc/source/internals.rst:
```diff
@@ -89,6 +89,25 @@ not check (or care) whether the levels themselves are sorted. Fortunately, the
+  ``ndarray`` or ``ExtensionArray``).
+
+So, for example, ``Series[category]._values`` is a ``Categorical``, while
+``Series[category]._ndarray_values`` is the underlying codes.
```
I think this section does not belong in the same document as "how to subclass", as this part is really internals for contributors to pandas. 
So I would maybe split this in two: external vs internal details, where the future documentation on how to use and define ExtensionArrays would also go into the external details part (maybe "Extending Pandas" would be a good title, with then information about the different strategies: accessor, extension array, subclassing, ..)

But that's for a separate PR (I can start with that), so for here the above is fine for me.

TomAugspurger replied:
`internal-internals.rst` and `external-internals.rst` ;)

I think that "How to subclass", and the eventual "extending pandas with custom array types" would be better in developer.rst, which is labeled as "This section will focus on downstream applications of pandas.".

jorisvandenbossche replied:
Ah, yes, I didn't see that the accessor documentation is actually already there (although, I personally don't find the parquet section that fitting in there, as it is not something typical you need to know when extending pandas. I can already start with moving it to the bottom of the file :-))

jreback commented on pandas/core/indexes/datetimes.py:
```diff
@@ -678,13 +678,36 @@ def _assert_tzawareness_compat(self, other):
 
+    @property
+    def _values(self):
+        # tz-naive -> ndarray
```
yeah we really ought to fix this

jreback reviewed (changes requested):
some small comments. ok to merge. But before adding *new* things, a refactor of the existing is in order to avoid all of this special casing. When this is out of sight (meaning not in a current PR), it is easy to forget the special casing. Maybe need a TODO detector . but already:
pandas) 

```
bash-3.2$ grep -r TODO pandas --include '*.py' |wc
     279    2233   22968
```
maybe put a special TODO(EA) or something to differentiate these

TomAugspurger committed:
Merge remote-tracking branch 'upstream/master' into index-values

TomAugspurger commented on pandas/core/base.py:
```diff
@@ -738,17 +739,17 @@ def data(self):
     def nbytes(self):
         """ return the number of bytes in the underlying data """
-        return self._values.nbytes
+        return self.values.nbytes
```
It depends on what we do with `PeriodIndex.values` and `IntervalIndex.values`.

If we change the public `.values` to be an ExtensionArray then there's no difference.

If we leave the public `.values` as an ndarray of object, then we prefer `._values` since it doesn't materialize the ndarray.

So far we've put off that decision, but we'll need to make a choice soon.

jorisvandenbossche replied:
You have to override this in DatetimeIndex (tz) then for now, this is the reason CI is failing (recursion error)

On the original comment: I think it is better for now to use `._values` instead of `.values` in such cases. `.values` is the public one that might not always be consistent for backwards compatibility, so let's internally use always `_values` to have something certainly consistent? (and if we decide `.values` to return extension arrays, it is just a public wrapper around `._values`, and if we decide not, we don't have to change anything internally)

TomAugspurger committed:
Cleanup

TomAugspurger commented:
Agreed entirely about the hackiness, especially in the set-ops. I should be able to do it in parallel with my Interval ExtensionArray PR.

TomAugspurger committed:
Override nbytes

TomAugspurger commented:
All green. Merging.

jorisvandenbossche commented:
Great!

jreback commented:
@TomAugspurger this broke the build. what is going on here?

jreback commented:
actually looks like just a network failure ok. But before merge anything else all of the other builds need to work.
